frame/support/src/storage/stream_iter.rs

```diff
@@ -0,0 +1,665 @@
+			let buffer_len = present_bytes + bytes_read;
+			unsafe {
+				self.buffer.set_len(buffer_len);
+			}
+
+			self.ensure_total_length_did_not_change(length_minus_offset)?;
+
+			self.offset += bytes_read as u32;
+
+			Ok(())
+		} else {
+			// The value was deleted, let's ensure we don't read anymore.
+			self.stop_reading();
+
+			Err("Value doesn't exist in the state?".into())
+		}
+	}
+
+	/// Returns if the value to read exists in the state.
+	fn exists(&self) -> bool {
+		self.exists
+	}
+
+	/// Reads directly into the given slice `into`.
+	///
+	/// Should be used when `into.len() > self.buffer.capacity()` to reduce the number of reads from
+	/// the state.
+	#[inline(never)]
+	fn read_big_item(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		let num_cached = self.buffer.len() - self.buffer_pos;
+
+		into[..num_cached].copy_from_slice(&self.buffer[self.buffer_pos..]);
+
+		self.buffer_pos = 0;
+		unsafe {
+			self.buffer.set_len(0);
+		}
+
+		if let Some(length_minus_offset) =
+			sp_io::storage::read(&self.key, &mut into[num_cached..], self.offset)
+		{
+			let required_to_read = (into.len() - num_cached) as u32;
+
+			if length_minus_offset < required_to_read {
+				return Err("Not enough data to fill the buffer".into())
+			}
+
+			self.ensure_total_length_did_not_change(length_minus_offset)?;
+
+			self.offset += required_to_read;
+
+			Ok(())
+		} else {
+			// The value was deleted, let's ensure we don't read anymore.
+			self.stop_reading();
+
+			Err("Value doesn't exist in the state?".into())
+		}
+	}
+
+	/// Ensures that the expected total length of the value did not change.
+	///
+	/// On error ensures that further reading is prohibited.
+	fn ensure_total_length_did_not_change(
+		&mut self,
+		length_minus_offset: u32,
+	) -> Result<(), codec::Error> {
+		if self.total_length == self.offset + length_minus_offset {
+			Ok(())
+		} else {
+			// The value total length changed, let's ensure we don't read anymore.
+			self.stop_reading();
+
+			Err("Storage value changed while it is being read!".into())
+		}
+	}
+
+	/// Ensure that we are stop reading from this value in the state.
+	///
+	/// Should be used when there happened an unrecoverable error while reading.
+	fn stop_reading(&mut self) {
+		self.offset = self.total_length;
+
+		self.buffer_pos = 0;
+		unsafe {
+			self.buffer.set_len(0);
+		}
+	}
+}
+
+impl codec::Input for StorageInput {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		Ok(Some(self.total_length.saturating_sub(
+			self.offset.saturating_sub((self.buffer.len() - self.buffer_pos) as u32),
+		) as usize))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		// If there is still data left to be read from the state.
+		if self.offset < self.total_length {
+			if into.len() > self.buffer.capacity() {
+				return self.read_big_item(into)
+			} else if self.buffer_pos + into.len() >= self.buffer.len() {
```

Comment:
This can be just `>`, right? (Since if they're equal the buffer should contain exactly the number of bytes we need to fulfill this read, so we don't need to fill it.)

Maybe to help readability we could also:
1. Add a `fn bytes_buffered(&self) -> usize { self.buffer.len() - self.buffer_pos }` helper.
2. Rewrite this as `into.len() > self.bytes_buffered()` to make it easier to understand.
3. Make the `into.len() + self.buffer_pos > self.buffer.len()` below also `into.len() > self.bytes_buffered()` (since it's checking exactly the same thing)
4. Use `self.bytes_buffered()` helper in three other places within this `impl` where `self.buffer.len() - self.buffer_pos` is used to make it more obvious what this expression is.